conda_index/index/sqlitecache.py

```diff
@@ -227,30 +252,46 @@ def extract_to_cache_unconditional(self, fn, abs_fn, size, mtime):
         }
 
         have = {}
-        package_stream = iter(package_streaming.stream_conda_info(abs_fn))
-        for tar, member in package_stream:
-            if member.name in wanted:
-                wanted.remove(member.name)
-                reader = tar.extractfile(member)
-                if reader is None:
-                    log.warn(f"{abs_fn}/{member.name} was not a regular file")
-                    continue
-                have[member.name] = reader.read()
-
-                # immediately parse index.json, decide whether we need icon
-                if member.name == INDEX_JSON_PATH:  # early exit when no icon
-                    index_json = json.loads(have[member.name])
-                    if index_json.get("icon") is None:
-                        wanted = wanted - {ICON_PATH}
-
-                if member.name in recipe_want_one:
-                    # convert yaml; don't look for any more recipe files
-                    have[member.name] = _cache_recipe(have[member.name])
-                    wanted = wanted - recipe_want_one
+        # second stream_conda_info "fileobj" parameter accepts Path or str
+        # inherited from ZipFile, bz2.open behavior, but we need to open the
+        # file ourselves.
+        with self.open(fn) as fileobj:
+            package_stream = iter(package_streaming.stream_conda_info(fn, fileobj))
+            for tar, member in package_stream:
+                if member.name in wanted:
+                    wanted.remove(member.name)
+                    reader = tar.extractfile(member)
+                    if reader is None:
+                        log.warn(f"{abs_fn}/{member.name} was not a regular file")
+                        continue
+                    have[member.name] = reader.read()
+
+                    # immediately parse index.json, decide whether we need icon
+                    if member.name == INDEX_JSON_PATH:  # early exit when no icon
+                        index_json = json.loads(have[member.name])
+                        if index_json.get("icon") is None:
+                            wanted = wanted - {ICON_PATH}
+
+                    if member.name in recipe_want_one:
+                        # convert yaml; don't look for any more recipe files
+                        have[member.name] = _cache_recipe(have[member.name])
+                        wanted = wanted - recipe_want_one
+
+                if not wanted:  # we got what we wanted
+                    package_stream.close()
+                    log.debug("%s early close", fn)
```

Comment:
The package stream isn't closed otherwise, is this a problem?

Reply:
It will close when we exhaust the iterator.